src/extensions/renderer/canvas/drawing-nodes.js

```diff
@@ -246,7 +258,57 @@ CRp.drawNode = function( context, node, shiftToOriginWithBb, drawLabel = true, s
       if( context.setLineDash ){ // for very outofdate browsers
         context.setLineDash( [ ] );
       }
+    }
+  };
+
+  let drawOutline = () => {
+    if( outlineWidth > 0 ){
+      // because outline and border are drawn along the same path,
+      // draw outline at border width plus twice the outline width
+      context.lineWidth = borderWidth + outlineWidth * 2;
+      context.lineCap = 'butt';
+
+      if( context.setLineDash ){ // for very outofdate browsers
+        switch( outlineStyle ){
+          case 'dotted':
+            context.setLineDash( [ 1, 1 ] );
+            break;
 
+          case 'dashed':
+            context.setLineDash( [ 4, 2 ] );
+            break;
+
+          case 'solid':
+            context.setLineDash( [ ] );
+            break;
+        }
+      }
+
+      drawPath();
+     
+      if( usePaths ){
+        context.stroke( path );
+      } else {
+        context.stroke();
+      }
+
+      const gco = context.globalCompositeOperation;
+      context.globalCompositeOperation = 'destination-out';
```

Comment:
Do we need to worry about what happens when an outline renders over another already rendered element here? Wouldn't it clip the outline above that too?

Reply:
I was thinking about that too the other day.  The GCO is fine for things like edge arrows, since they're so small.  However, there are important use cases for having semitransparent nodes -- e.g. with overlap -- and the GCO would visually delete things.

It would be best to use a clip operation here.  It can be a bit cheaper if you use the `nodePathCache` for the clipping rather than having to redraw/regenerate the node body.

Reply:
I've managed to update the rendering to clip instead of use gco, but only by rendering a rect with the node shape cut out, and clipping to that. I might very well be missing something, but clipping to the node shape has the opposite effect of clipping the outer portion of the outline instead of the inside portion. 

There are still a bunch of irregular shapes whose outline exceeds the node width/height + outline width though, and I'm not sure how to accurately expand the rect bounds in those cases, and when calculating the overall node bounds in `bounds.js` if an outline is present. I can apply an arbitrary multiplier, but that feels like a hack which would likely have adverse implications on node layout, etc.

Reply:
That's a good point.  Here are some initial thoughts:

(1) Two multipliers, one for width and one for height, might suffice for predefined shapes.  One possibility is something like `bb.w = k_w * outlineWidth + otherMeasuresForWidth`.

(2) Assuming we go with (1), how would we address custom `polygon` shapes?  This leads into (3).  Alternatively, we could lean on the existing `bounds-expansion` property.  The dev, in that case, must specify their own expansion of the bounds to accommodate the outline on their custom polygon.

(3) The proper way to calculate the adjusted bounding box would be to take the intersection of the relevant lines in the shape, when they're pushed out by the outline width.  This would be more expensive, and it would mean that we may as well use a path for the outline, since we'd have the points to form the path anyway.

Reply:
Taking a step back, pretty much all of the complexity that's arisen has come with trying to solve the use case of the background and/or border having a partial opacity, and not wanting any of the outline below to show through. 

But we already have that with borders, where a border with partial opacity effectively creates _two_ visual borders due to the fact that the inner half overlaps the node shape:

<img width="593" alt="Screen Shot 2023-09-27 at 11 14 10 AM" src="https://github.com/cytoscape/cytoscape.js/assets/1434577/0f138429-112e-4473-8cf0-d6a1744977d8">

In an ideal world, it might be worth considering rendering both borders and outlines such that they never overlapped other elements. In practice though this has proven to be less straightforward than it would seem, whether accomplished by offsetting the stroke or by clipping, and there's a chance that the computational cost would end up outweighing the aesthetics, making it hard to justify the tradeoff. 

While adding outlines admittedly moves us closer to that potential decision point, I wonder if we could/should consider simplifying the outline implementation for now by not worrying about clipping the inner portion, and consider that broader rethink around how strokes are rendered (whether border or outline) down the road.

Reply:
What if we rebrand the feature?

What you're describing sounds like a `under-border` rather than an `outline`.  That way, you can still use it as an outline -- with the opacity caveats etc. -- but it would have clear expectations for other devs who use the feature.  Nobody would be disappointed about the transparent node use cases, since it's obvious that those use cases aren't suited to an under-border.

The implementation would be identical to the existing borders, apart from the draw order.

Reply:
@maxkfranz that could be a possibility as a fallback, but my preference would be to keep the outline framing if possible. I will give the suggestion you proposed above (calculating the path by pushing out the points by the outline width) a shot.

Reply:
Sounds good